liferay-appsec/liferay-portal
# LPS-194700 As a Developer I want to be able to list every inactive User via a Headless-API endpoint

Motivation
=======
https://liferay.atlassian.net/browse/LPS-188650

Proposed Solution
========
Two GET endpoints have been added to the headless-admin-user application:

- `/headless-admin-user/v1.0/user-accounts/by-status/${status}`
where status is either "active" or "inactive", to allow getting users by its status value.

- `/headless-admin-user/v1.0/user-accounts/by-email-address/${emailAddress}` to allow getting a user by its email address, regardless of its current status. 

An additional **readOnly** field `status` has been added to the `UserAccount` entity (with value `Active` or `Inactive`), which is returned in every endpoint returning a `UserAccount` or a collection of `UserAccount`s.

The story mentions the response should include the following fields:
`userID`
`firstName`
`lastName`
`emailAddress`
`Role`
`Status(active/inactive)`

Besides `status` (which has been added) and `emailAddress`, the names of the fields are already defined in the UserAccount entity with slightly different names:

`userID -> id`
`firstName -> givenName`
`lastName -> familyName`
`Role -> roleBriefs`

The new endpoint returns instances of the entity `UserAccount`, and therefore contains all its fields with the already existing field names.

Steps to Verify
========
- Make sure you have enabled feature flag LPS-188420
- Create several users via UI or headless API
- Deactivate some of them via UI

- Invoke the new endpoint `/headless-admin-user/v1.0/user-accounts/by-status/inactive` and verify that only the inactive users are returned
- Verify that users returned contain field `"status" : "Inactive"`

- Invoke the new endpoint `/headless-admin-user/v1.0/user-accounts/by-status/active` and verify that only the active users are returned
- Verify that users returned contain field `"status" : "Active"`

- Invoke the new endpoint `/headless-admin-user/v1.0/user-accounts/by-email-address/${emailAddress}` for an inactive user and verify that the inactive user is returned
- Verify that user returned contains field `"status" : "Inactive"`

- Invoke the new endpoint `/headless-admin-user/v1.0/user-accounts/by-email-address/${emailAddress}` for an active user and verify that the active user is returned
- Verify that user returned contains field `"status" : "Active"`

- Verify that filter, page, search and sort can be included as parameters to `/headless-admin-user/v1.0/user-accounts/by-status/${status}` similarly to how they are included for the already existing endpoint `/headless-admin-user/v1.0/user-accounts` 

References to existing code
========
The `/user-accounts/by-status/${status}` is the endpoint chosen to display users with a specified state.
This follows the pattern of already existing endpoints like:
- `/sites/{siteId}/data-definitions/by-content-type/{contentType}`
- `/structured-contents/{structuredContentId}/by-version/{version}`
while making the api flexible enough for supporting any status value that may be supported in the future without having to add additional endpoints.

//cc @ealonso @jkappler
